bpf: Do not bypass conntrack if running kube-proxy+vxlan+endpoint routes

pchaigno opened the pull request:
This commit fixes a broken packet path on `pod --> remote pod` via VIP when running with encapsulation and per-endpoint routes, with kube-proxy handling DNATing. The following details the broken packet path:

- SYN leaves lxc device for stack with `pod1IP -> VIP`
- SYN goes through cilium_host and cilium_net with `pod1IP -> pod2IP`
- SYN is sent to second node through tunnel with `pod1IP -> pod2IP`
- SYN+ACK arrives from tunnel on `cilium_vxlan` with `pod2IP -> pod1IP`
- SYN+ACK is processed by `ipvX_local_delivery` and redirected in BPF directly to the lxc device
- SYN+ACK arrives on lxc device with `pod2IP -> pod1IP`
- RST is sent because pod2IP doesn't match VIP

Because we use `redirect()` in BPF, the SYN+ACK packet bypasses most of the stack (including conntrack) on its way from `cilium_vxlan` to the lxc device. By bypassing conntrack, we also bypass the reverse DNAT.

The bug doesn't occur when per-endpoint routes are disabled because, in that case, the SYN is SNATed to the `cilium_host` IP before leaving through the tunnel. The SYN+ACK packet therefore take a different path and goes through conntrack.

Fixes: https://github.com/cilium/cilium/issues/14646
Fixes: https://github.com/cilium/cilium/issues/14929
Co-authored-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>

```release-note
Fix remote pod connectivity through VIP in tunneling mode with kube-proxy and per-endpoint routes.
Fix IPv6 connectivity to BPF HostPort when kube-proxy is installed 
```

pchaigno committed:
bpf: Do not bypass conntrack if running kube-proxy+vxlan+endpoint routes

This commit fixes a broken packet path on pod --> remote pod via VIP
when running with encapsulation and per-endpoint routes, with kube-proxy
handling DNATing. The following details the broken packet path:

- SYN leaves lxc device for stack             with pod1IP -> VIP
- SYN goes through cilium_host and cilium_net with pod1IP -> pod2IP
- SYN is sent to second node through tunnel   with pod1IP -> pod2IP
- SYN+ACK arrives from tunnel on cilium_vxlan with pod2IP -> pod1IP
- SYN+ACK is processed by ipvX_local_delivery
  and redirected in BPF directly to the lxc
  device
- SYN+ACK arrives on lxc device               with pod2IP -> pod1IP
- RST is sent because pod2IP doesn't match VIP

Because we use redirect() in BPF, the SYN+ACK packet bypasses most of
the stack (including conntrack) on its way from cilium_vxlan to the lxc
device. By bypassing conntrack, we also bypass the reverse DNAT.

The bug doesn't occur when per-endpoint routes are disabled because, in
that case, the SYN is SNATed to the cilium_host IP before leaving
through the tunnel. The SYN+ACK packet therefore takes a different path
and goes through conntrack.

Fixes: https://github.com/cilium/cilium/issues/14646
Co-authored-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet reviewed (approved):
Looks good to me, thanks!
(The checkpatch issue reported here has been disabled in a later PR, it can be safely ignored in the check below)

brb commented on bpf/bpf_lxc.c:
```diff
@@ -1243,9 +1257,14 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 			   LXC_ID, ifindex, *reason, monitor);
 
+#if !defined(ENABLE_ROUTING) && defined(ENCAP_IFINDEX) && !defined(ENABLE_NODEPORT)
+	/* See comment in IPv6 path. */
```
Nit: I think a consensus is to document IPv4 path, and refer IPv6 to IPv4 for comments.

brb reviewed (commented):
I think by VIP in the commit description you meant NodePort/Loadbalancer/exteranlIP service? Just noting, that ClusterIP services should be handled by bpf_lxc when kube-proxy is in use.

brb commented on bpf/bpf_lxc.c:
```diff
@@ -999,9 +1000,21 @@ ipv6_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
+	 * will drop them.
+	 * See #14646 for details.
+	 */
+	ctx_change_type(ctx, PACKET_HOST);
```
This means that the packet will enter `ipvX_policy()` twice?

pchaigno replied:
Not as far as I can see. The packet will still leave for the stack. What makes you think it would be processed twice?

brb replied:
My thinking was:

1. `ipvX_policy()` is invoked from bpf_overlay for a reply.
2. The reply is passed to stack, and the reverse DNAT translation is performed.
3. (This is where I can be wrong) The route lookup passes the reply to the veth which has bpf_lxc attached which executes `ipvX_policy()` again.

pchaigno replied:
Ah, I see. I don't think the current patch will change that. It's already happening without this patch: https://github.com/cilium/cilium/issues/14657.

I haven't had time to look at the above issue yet, but one of the two policy enforcements (from bpf_overlay or from bpf_lxc) likely should be compiled out and isn't.

brb replied:
It's not important, but the current patch should change the path because `return redirect_ep(ifindex, from_host);` won't happen.

joestringer replied:
@brb @pchaigno is the feedback here clear / addressed? Seems like this is the only open discussion topic for this 1.9-blocker.

brb replied:
This was non-blocking comment from my side. 

> the current patch should change the path

I just want our understanding on this to be on the same page.

pchaigno replied:
Yes, the current patch will change the path of the packet to go through the stack instead of taking the `redirect()` shortcut. I do not think however that the packet path impacts #14657; the packet is already processed twice and will continue to be with this patch. We should fix #14657 in a separate PR.

brb reviewed (approved):
:rocket:

pchaigno committed:
iptables: Populate CILIUM_FORWARD chain for IPv6

Kube-proxy installs the following rule in the iptables FORWARD chain:

    -A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP

That rule causes iptables to drop packets whenever we have a path
asymmetry with for example a SYN packet going through iptables and a
SYN+ACK packet skipping iptables via BPF's redirect(). In that example,
the next ACK packet would be dropped by that rule because the connection
is invalid (SYN+ACK is missing) from iptables' point of view.

To avoid such drops, in the IPv4 case, Cilium installs rules in the
FORWARD chain (via its own CILIUM_FORWARD chain) to bypass the
subsequent -j DROP rule from kube-proxy. For example, with per-endpoint
routes enabled:

    -A FORWARD -m comment --comment "cilium-feeder: CILIUM_FORWARD" -j CILIUM_FORWARD
    -A CILIUM_FORWARD -o cilium_host -m comment --comment "cilium: any->cluster on cilium_host forward accept" -j ACCEPT
    -A CILIUM_FORWARD -i cilium_host -m comment --comment "cilium: cluster->any on cilium_host forward accept (nodeport)" -j ACCEPT
    -A CILIUM_FORWARD -i lxc+ -m comment --comment "cilium: cluster->any on lxc+ forward accept" -j ACCEPT
    -A CILIUM_FORWARD -i cilium_net -m comment --comment "cilium: cluster->any on cilium_net forward accept (nodeport)" -j ACCEPT
    -A CILIUM_FORWARD -o lxc+ -m comment --comment "cilium: any->cluster on lxc+ forward accept" -j ACCEPT
    -A CILIUM_FORWARD -i lxc+ -m comment --comment "cilium: cluster->any on lxc+ forward accept (nodeport)" -j ACCEPT

The same rules are however missing from the IPv6 tables. This commit
fixes this inconsistency.

Signed-off-by: Paul Chaignon <paul@cilium.io>
